### PR TITLE
vim: Update to 9.1.0516 and variants

### PR DESCRIPTION
**Pull request by kaimingguo**
- Add `python 3.12` variant.
- Remove `>= 18` Ruby minimum required version for Vim build.
- Add `ruby 3.3` support, and the default variants.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 12.7.5 21H1222 x86_64
Xcode 14.2 14C18

###### Verification 

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: editors/vim/Portfile [hunk 1]
```diff
@@ -138,51 +138,81 @@ foreach s ${pythons_suffixes} {
     "
 }
 
-variant ruby requires ruby18 description {Compatibility variant, requires +ruby18} {}
-variant ruby18 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby30 description {Enable Ruby scripting} {
+variant ruby requires ruby33 description {Compatibility variant, requires +ruby33} {}
```
Comment (raimue) [inline editors/vim/Portfile:141]: I would suggest dropping this +ruby variant. We also used to have a +python, but it only made it more complicated for users to switch to a newer variant on upgrade.

File: editors/vim/Portfile [hunk 2]
```diff
@@ -138,51 +138,81 @@ foreach s ${pythons_suffixes} {
...
+variant ruby19 conflicts ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby26 ruby27 ruby30 ruby31 ruby32 ruby33 description {Enable Ruby scripting} {}
```
Comment (raimue) [inline editors/vim/Portfile:142]: Would it be possible to solve this with a loop as for the +pythonXY variants? It seems difficult to maintain the conflicting versions like this.

File: editors/vim/Portfile [hunk 3]
```diff
@@ -138,51 +138,81 @@ foreach s ${pythons_suffixes} {
...
+variant ruby25 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby26 ruby27 ruby30 ruby31 ruby32 ruby33 description {Enable Ruby scripting} {}
+variant ruby26 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby27 ruby30 ruby31 ruby32 ruby33 description {Enable Ruby scripting} {}
+variant ruby27 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby26 ruby30 ruby31 ruby32 ruby33 description {Enable Ruby scripting} {}
+variant ruby30 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby26 ruby27 ruby31 ruby32 ruby33 description {Enable Ruby scripting} {}
+variant ruby31 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby26 ruby27 ruby30 ruby32 ruby33 description {Enable Ruby scripting} {}
+variant ruby32 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby26 ruby27 ruby30 ruby31 ruby33 description {Enable Ruby scripting} {}
+variant ruby33 conflicts ruby19 ruby20 ruby21 ruby22 ruby23 ruby24 ruby25 ruby26 ruby27 ruby30 ruby31 ruby32 description {Enable Ruby scripting} {}
+
+if {[variant_isset ruby33]} {
```
Comment (raimue) [inline editors/vim/Portfile:156]: Why do you use `variant_isset` instead of putting it into the variant body as it has been before?

**Review by raimue — CHANGES_REQUESTED** [review]
Thank you for the port update. However, it cannot be merged like this directly. Please respect the comment regarding the vim version:

> # The vim port should always follow the patch level used in the MacVim port. This
> # is because they both share the same set of configuration files and ensures
> # that the user's configuration files will always work with both.

vim 9.1.0 is the latest version shipped by the MacVim port. This synchronization also extends to other parts of the port like the feature selection. The changes to the Ruby variants should also be applied to MacVim in the same way.

Is the new vim version required for the new +ruby33 variant? Could we apply that without also updating the version?

File: editors/vim/Portfile (hunk 1 above)
Comment (kaimingguo) [reply editors/vim/Portfile:141]: I agree this about not presetting a version. I'll be validating `MacVim` port and deleting the variant.

File: editors/vim/Portfile (hunk 3 above)
Comment (kaimingguo) [reply editors/vim/Portfile:156]: Separate version selection conflicts, and use `variant_isset` for logical separation. Keeping scripts clean.

**Comment by kaimingguo** [comment]
> Thank you for the port update. However, it cannot be merged like this directly. Please respect the comment regarding the vim version:
> 
> > # The vim port should always follow the patch level used in the MacVim port. This
> > # is because they both share the same set of configuration files and ensures
> > # that the user's configuration files will always work with both.
> 
> vim 9.1.0 is the latest version shipped by the MacVim port. This synchronization also extends to other parts of the port like the feature selection. The changes to the Ruby variants should also be applied to MacVim in the same way.
> 
> Is the new vim version required for the new +ruby33 variant? Could we apply that without also updating the version?

Okay, I'll follow the patch level used in the MacVim. The new `vim` version is required for the Ruby 1.9 or newer.

**Pull request closed without merging**